auth/auth.go

```diff
@@ -417,110 +418,116 @@ func (auth *Authenticator) AuthenticateUser(username string, password string) Us
+			return nil, time.Time{}, err
+		}
+
+		// Extract issuer and audience(s) from JSON Web Token.
+		issuer, audiences, err := GetIssuerWithAudience(jwt)
+		base.Debugf(base.KeyAuth, "JWT issuer: %v, audiences: %v", base.UD(issuer), base.UD(audiences))
+		if err != nil {
+			base.Debugf(base.KeyAuth, "Error getting issuer and audience from token: %v", err)
+			return nil, time.Time{}, err
+		}
 
-	base.Debugf(base.KeyAuth, "Call GetProviderForIssuer w/ providers: %+v", base.UD(providers))
-	provider := providers.GetProviderForIssuer(issuer, audiences)
-	base.Debugf(base.KeyAuth, "Provider for issuer: %+v", base.UD(provider))
+		base.Debugf(base.KeyAuth, "Call GetProviderForIssuer w/ providers: %+v", base.UD(providers))
+		provider := providers.GetProviderForIssuer(issuer, audiences)
+		base.Debugf(base.KeyAuth, "Provider for issuer: %+v", base.UD(provider))
+	}
 
 	if provider == nil {
-		return nil, jose.JWT{}, base.RedactErrorf("No provider found for issuer %v", base.UD(issuer))
+		return nil, time.Time{}, base.RedactErrorf("No provider found for issuer %v", base.UD(issuer))
 	}
 
-	// VerifyJWT validates the claims and signature on the JWT
+	// Get client for issuer
 	client := provider.GetClient(callbackURLFunc)
 	if client == nil {
-		return nil, jose.JWT{}, fmt.Errorf("OIDC client was not initialized")
+		return nil, time.Time{}, fmt.Errorf("OIDC client was not initialized")
 	}
 
-	err = client.VerifyJWT(jwt)
+	// VerifyJWT validates the claims and signature on the JWT
+	idToken, err := client.VerifyJWT(token)
 	if err != nil {
 		base.Debugf(base.KeyAuth, "Client %v could not verify JWT. Error: %v", base.UD(client), err)
-		return nil, jwt, err
+		return nil, time.Time{}, err
 	}
 
-	return auth.authenticateJWT(jwt, provider)
+	identity, err := GetIdentity(idToken)
+	if err != nil {
+		base.Debugf(base.KeyAuth, "Error getting identity from token", base.UD(identity), err)
+	}
+	return auth.authenticateOIDCIdentity(identity, provider)
 }
 
 // Authenticates a user based on a JWT token obtained directly from a provider (auth code flow, refresh flow).
 // Verifies the token claims, but doesn't require signature verification.
 // If the token is validated but the user for the username defined in the subject claim doesn't exist,
 // creates the user when autoRegister=true.
-func (auth *Authenticator) AuthenticateTrustedJWT(token string, provider *OIDCProvider, callbackURLFunc OIDCCallbackURLFunc) (User, jose.JWT, error) {
+func (auth *Authenticator) AuthenticateTrustedJWT(token string, provider *OIDCProvider, callbackURLFunc OIDCCallbackURLFunc) (User, time.Time, error) {
+	base.Debugf(base.KeyAuth, "AuthenticateTrustedJWT called with token: %s", base.UD(token))
 
-	// Parse JWT
-	jwt, err := jose.ParseJWT(token)
-	if err != nil {
-		base.Debugf(base.KeyAuth, "Error parsing JWT in AuthenticateTrustedJWT: %v", err)
-		return nil, jose.JWT{}, err
+	if provider.ClientID == nil {
```

Comment:
Makes much sense. Refactored clientID definition from *string to string, added changes to check for empty string instead of nil in NewDatabaseContext, and removed the duplicate check from AuthenticateTrustedJWT.